### PR TITLE
[GOBBLIN-1947] Send workUnitChange event when helix task consistently fail

### PR DESCRIPTION
**Pull request by hanghangliu**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-1947] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1947

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):
When YarnAutoScalingManager detect helix task consistently fail, give an option to send WorkUnitChangeEvent to let GobblinHelixJobLauncher handle the event and split the work unit during runtime. This can help resolving consistent failing containers issue(like OOM) during runtime instead of relying on replaner to restart the whole pipeline

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Updated test cases and tested in cluster

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java [hunk 1]
```diff
@@ -275,17 +359,18 @@ protected void executeCancellation() {
...
                                                                                     RetryException {
     String jobName = this.jobContext.getJobId();
     try (ParallelRunner stateSerDeRunner = new ParallelRunner(this.stateSerDeRunnerThreads, this.fs)) {
-      for (String workUnitId : workUnitIdsToRemove) {
+      for (String helixTaskId : helixTaskIdsToRemove) {
+        String workUnitId = helixIdTaskConfigMap.get(helixTaskId).getConfigMap().get(ConfigurationKeys.TASK_ID_KEY);
         taskRetryer.call(new Callable<Boolean>() {
           @Override
           public Boolean call() throws Exception {
             String taskId = workUnitToHelixConfig.get(workUnitId).getId();
```
Comment (ZihanLi58) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java:371]: do you still need this?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java [hunk 2]
```diff
@@ -514,6 +599,7 @@ private TaskConfig getTaskConfig(WorkUnit workUnit, ParallelRunner stateSerDeRun
     rawConfigMap.put(GobblinClusterConfigurationKeys.TASK_SUCCESS_OPTIONAL_KEY, "true");
     TaskConfig taskConfig = TaskConfig.Builder.from(rawConfigMap);
     workUnitToHelixConfig.put(workUnit.getId(), taskConfig);
```
Comment (ZihanLi58) [inline gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java:601]: same here, do we still need this? I feel you want to use helix TaskId instead of work unit Id here

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java [hunk 3]
```diff
@@ -507,9 +513,10 @@ public void apply(JobListener jobListener, JobContext jobContext)
         }
 
         // Perform work needed before writing is done
-        Boolean canCleanUp = this.canCleanStagingData(this.jobContext.getJobState());
-        workUnitStream = closer.register(new DestinationDatasetHandlerService(jobState, canCleanUp, this.eventSubmitter))
-            .executeHandlers(workUnitStream);
+        this.canCleanUpStagingData = this.canCleanStagingData(this.jobContext.getJobState());
```
Comment (ZihanLi58) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java:516]: Is this from other change? seems not related to this PR

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java (hunk 2 above)
Comment (ZihanLi58) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java:601]: Basically seems helixIdTaskConfigMap and workUnitToHelixConfig are similar, can we just use one of them to reduce complexity?

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java [hunk 4]
```diff
@@ -254,6 +270,74 @@ protected void runWorkUnits(List<WorkUnit> workUnits) throws Exception {
...
+  @Override
+  @Subscribe
+  public void handleWorkUnitChangeEvent(WorkUnitChangeEvent workUnitChangeEvent)
+      throws InvocationTargetException {
+    log.info("Received WorkUnitChangeEvent with old Task {} and new WU {}",
+        workUnitChangeEvent.getOldTaskIds(), workUnitChangeEvent.getNewWorkUnits());
+    final JobState jobState = this.jobContext.getJobState();
+    List<WorkUnit> workUnits = workUnitChangeEvent.getNewWorkUnits();
+    // Use old task Id to recalculate new work units
+    if(workUnits == null || workUnits.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (hanghangliu) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java:291]: When we send workUnitChangeEvent from YarnAutoScalingManager. The yarn service only has the information of helix, so may not easy to pre-calculate the new workUnit as this process need KafkaSource, which yarn isn't aware of.

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java [hunk 5]
```diff
@@ -254,6 +270,74 @@ protected void runWorkUnits(List<WorkUnit> workUnits) throws Exception {
...
+    workUnitStream = this.executeHandlers(workUnitStream, this.destDatasetHandlerService);
+    this.processWorkUnitStream(workUnitStream, jobState);
+    try {
+      this.removeTasksFromCurrentJob(workUnitChangeEvent.getOldTaskIds());
+      this.addTasksToCurrentJob(workUnits);
+    } catch (Exception e) {
+      //todo: emit some event to indicate there is an error handling this event that may cause starvation
+      log.error("Failed to process WorkUnitChangeEvent with old tasks {} and new workunits {}.",
+          workUnitChangeEvent.getOldTaskIds(), workUnits, e);
+      throw new InvocationTargetException(e);
```
[the remark this answers is not in the document]
Comment (hanghangliu) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java:312]: Throwing error actually won't fail the application, so we rely on the Retryer and replanner(if Retryer also failed) here. I've run the test for a week and didn't see any error throwing, but I do agree this may be hard to debug. 
I've tried to restart the whole workflow but it's not very straightforward

[the remark this answers is not in the document]
Comment (hanghangliu) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java:312]: For the InvocationTargetException, it's actually inherited from [the super class](https://github.com/apache/gobblin/blob/694ed37649ce4d48e4f4810287c7245339d19069/gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java#L281) which written by you :)

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java (hunk 1 above)
Comment (hanghangliu) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java]: removed

File: gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java (hunk 2 above)
Comment (hanghangliu) [reply gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java]: removed workUnitToHelixConfig

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java (hunk 3 above)
Comment (hanghangliu) [reply gobblin-runtime/src/main/java/org/apache/gobblin/runtime/AbstractJobLauncher.java:520]: this is actually needed and may easily get overlooked. It's a process that make sure the workUnit can handle shards for target directory. Took me a while the figure out and debug when testing in the cluster...

**Review by ZihanLi58 — APPROVED** [review]
+1
